cpp/src/cider/common/benchmark/HashMapBenchmark.cpp

```diff
@@ -0,0 +1,168 @@
+      ++it->getMapped();
+  }
+
+  for (auto value : data) {
+    auto it = map.find(value);
+    auto curr = ++it;
+    if (curr)
+      curr->getMapped();
+  }
+  watch.stop();
+  std::cerr << std::fixed << std::setprecision(2) << "HashMap (" << name
+            << "). Size: " << map.size() << ", elapsed: " << watch.elapsedSeconds()
+            << " (" << data.size() / watch.elapsedSeconds() << " elem/sec.)"
+            << ", collisions: " << map.getCollisions() << std::endl;
+}
+
+// template <typename Map>
+// void insert(Map& map, StringRef& k) {
+//   bool inserted;
+//   typename Map::LookupResult it;
+//   map.emplace(k, it, inserted, nullptr);
+//   if (inserted)
+//     it->getMapped() = 1;
+//   else
+//     ++it->getMapped();
+//   std::cout << map.find(k)->getMapped() << std::endl;
+// }
+
+template <typename T>
+static std::tuple<std::vector<T>, std::vector<bool>> generateAndFillVector(
+    const size_t row_num,
+    const GeneratePattern pattern,
+    const int32_t
+        null_chance,  // Null chance for each column, -1 represents for unnullable
+                      // column, 0 represents for nullable column but all data is not
+                      // null, 1 represents for all rows are null, values >= 2 means
+                      // each row has 1/x chance to be null.
+    const T value_min = N_MIN,
+    const T value_max = N_MAX) {
+  std::vector<T> col_data(row_num);
+  std::vector<bool> null_data(row_num);
+  std::mt19937 rng(std::random_device{}());  // NOLINT
+  switch (pattern) {
+    case GeneratePattern::Sequence:
+      for (auto i = 0; i < row_num; ++i) {
+        null_data[i] = Random::oneIn(null_chance, rng) ? (col_data[i] = N_MIN, true)
+                                                       : (col_data[i] = i, false);
+      }
+      break;
+    case GeneratePattern::Random:
+      if (std::is_integral<T>::value) {
+        for (auto i = 0; i < col_data.size(); ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng)
+                             ? (col_data[i] = N_MIN, true)
+                             : (col_data[i] = static_cast<T>(
+                                    Random::randInt64(value_min, value_max, rng)),
+                                false);
+        }
+      } else if (std::is_floating_point<T>::value) {
+        for (auto i = 0; i < col_data.size(); ++i) {
+          null_data[i] = Random::oneIn(null_chance, rng)
+                             ? (col_data[i] = N_MIN, true)
+                             : (col_data[i] = static_cast<T>(
+                                    Random::randFloat(value_min, value_max, rng)),
+                                false);
+        }
+      } else {
+        std::string str = "Unexpected type:";
+        str.append(typeid(T).name()).append(", could not generate data.");
+        CIDER_THROW(CiderCompileException, str);
+      }
+      break;
+  }
+  return std::make_tuple(col_data, null_data);
+}
+
+int main(int argc, char** argv) {

```

Comment:
Please use a benchmark framework instead.

Reply:
Changed to google benchmark framework.